examples/factory/tests/tests.rs

```diff
@@ -0,0 +1,231 @@
+//! Tests for the factory smart contract.
+use concordium_smart_contract_testing::*;
+use concordium_std::{Cursor, Deserial};
+use factory_smart_contract::product::{Product, ProductState};
+
+/// The test accounts.
+const ALICE: AccountAddress = AccountAddress([0; 32]);
+const BOB: AccountAddress = AccountAddress([1; 32]);
+
+const SIGNER: Signer = Signer::with_one_key();
+const ACC_INITIAL_BALANCE: Amount = Amount::from_ccd(10000);
+
+/// Test the case of producing two products from one factory.
+#[test]
+fn test_success() {
+    let (mut chain, module_ref) = initialize_chain();
+    // Set up the factory instance.
+    let factory = create_factory(&mut chain, module_ref);
+
+    // Create a product instance.
+    let product = create_product(&mut chain, module_ref, ALICE);
+    // Check that it is uninitialized
+    let prod_state = view_product(&chain, product).expect("Get view of uninitialized product");
+    assert_eq!(prod_state, ProductState::Uninitialized);
+    // "Produce" the product with the factory.
+    call_produce(&mut chain, factory, product, ALICE).expect("Produce successful");
+    // Check that it is initialized as we expect.
+    let prod_state = view_product(&chain, product).expect("Get view of initialized product");
+    assert_eq!(
+        prod_state,
+        ProductState::Initialized(Product {
+            factory,
+            index: 0,
+        })
+    );
+
+    // Create and produce a second product (from a different account), and check the
+    // result.
+    let product2 = create_product(&mut chain, module_ref, BOB);
+    call_produce(&mut chain, factory, product2, BOB).expect("Produce successful");
+    let prod_state = view_product(&chain, product2).expect("Get view of initialized product");
+    assert_eq!(
+        prod_state,
+        ProductState::Initialized(Product {
+            factory,
+            index: 1,
+        })
+    );
+
+    // Check that the factory records having 2 products, and that they match what we
+    // expect.
+    assert_eq!(view_product_count(&chain, factory).expect("View product count successful"), 2);
+    assert_eq!(view_product_at(&chain, factory, 0).expect("View of product at index"), product);
+    assert_eq!(view_product_at(&chain, factory, 1).expect("View of product at index"), product2);
+}
+
+/// Test that a product instance cannot be produced using an account that did
+/// not create the product instance.
+#[test]
+fn test_not_authorized() {
+    let (mut chain, module_ref) = initialize_chain();
+    let factory = create_factory(&mut chain, module_ref);
+    // ALICE creates the product instance, but BOB invokes produce on the factory.
+    // This should fail.
+    let product = create_product(&mut chain, module_ref, ALICE);
+    call_produce(&mut chain, factory, product, BOB).expect_err(
```

Comment:
It would be good to check if the correct error is produced and not only some error/revert in the smart contract happened.

Similar in other test cases.